sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1098,6 +1106,87 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
     }
   }
 
+  /**
+   * Add an [[Unpivot]] to a logical plan.
+   */
+  private def withUnpivot(
+      ctx: UnpivotClauseContext,
+      query: LogicalPlan): LogicalPlan = withOrigin(ctx) {
+    // this is needed to create unpivot and to filter unpivot for nulls further down
+    val valueColumnNames =
+      Option(ctx.unpivotOperator().unpivotSingleValueColumnClause())
+        .map(_.unpivotValueColumn().identifier().getText)
+        .map(Seq(_))
+      .getOrElse(
+        Option(ctx.unpivotOperator().unpivotMultiValueColumnClause())
+          .map(_.unpivotValueColumns.asScala.map(_.identifier().getText).toSeq)
+          .get
+      )
+
+    val unpivot = if (ctx.unpivotOperator().unpivotSingleValueColumnClause() != null) {
+      val unpivotClause = ctx.unpivotOperator().unpivotSingleValueColumnClause()
+      val variableColumnName = unpivotClause.unpivotNameColumn().identifier().getText
+      val unpivotColumns = unpivotClause.unpivotColumns.asScala.map(visitUnpivotColumn).toSeq
+
+      Unpivot(
+        None,
+        Some(unpivotColumns.map(Seq(_))),
```

Comment:
How can we make sure these value columns will end up with `Attribute`?

Reply:
Method `visitUnpivotColumn` returns an `UnresolvedAttribute`, which turns into an `AttributeReference`. Otherwise the analyser fails.

Reply:
`SELECT a.b.c` also produce `UnresolvedAttribute`, and it may end up with column `c` in table `b` under database `a` (thus it's `AttributeReference`), or nested column `c` of column `b` in table `a` (thus it's `GetStructField`)

Reply:
I am not sure what we want to support here:

```
SELECT *
FROM a.b
UNPIVOT (
  val for col in (c)
)
```
when `a` is a database and `b` is a table.

```
SELECT *
FROM a.b.*
UNPIVOT (
  val for col in (c)
)
```
when `a` is a table and `b` is a struct.

In either way, `c` is an `AttributeReference`, right?

Reply:
I'm saying that it's impossible to enforce `AttributeReference` in the parser side. `UnresolvedAttribute` can be resolved to `AttributeReference` or `GetStructField`. Please refer to the rule `ResolveReferences`. We need to enforce `AttributeReference` in the analyzer rule.

Reply:
I am still missing why we do not want to allow all `AttributeReference`s here, why shall we exclude `GetStructField` and any of the other `ExtractValue`s here?

Reply:
Done in 70f986a.

I think the error message needs some improvement, it is likely not clear to users what is meant with `Attribute`:

    "UNPIVOT requires given {given} to be Attributes when no {empty} are given: [<types>]"

Maybe `column names` is better.

Reply:
@cloud-fan I'd say, I have resolved all comments. And I have rebased with latest master.